### ISSUE TITLE
Plugin order not respected

### ISSUE THREAD

**Issue by giusepetroso** [postcss/postcss#1786]
Hi guys,

I don't know if I am doing something wrong but I'm experiencing some problem figuring out why plugins seems to run not in the order I specified in postcss.config.js.

I created a new vue 3 project and installed **tailwindcss**, **postcss** and **some plugins** to show what I am talking about, I will try to explain some steps I'm doing to figure out.
I published it on github [here](https://github.com/giusepetroso/postcss-plugin-issue)

Starting with **main.css**
![image](https://user-images.githubusercontent.com/17656277/194708506-352c3ffe-998c-4190-aeb3-8acce3b9d7b8.png)
- Importing tailwind (with @import because if I put @tailwind directives before import postcss complains it)
- Importing colors.css file with some css vars and nothing else
![image](https://user-images.githubusercontent.com/17656277/194708679-07234cc7-9094-40f8-a6e9-d0b02973ae39.png)
- Importing components/badge/index.css which imports a mixin, applies some tailwind rules and uses @mixin inside a @each 
![image](https://user-images.githubusercontent.com/17656277/194709483-a32633ea-31d8-45ea-81a9-06c9e274e4d7.png)
![image](https://user-images.githubusercontent.com/17656277/194709494-8f7097ca-4ea7-4947-b660-ceb09afc2764.png)

The **package.json**
![image](https://user-images.githubusercontent.com/17656277/194708872-ab67b129-218f-49e8-b1e3-d71b87e894fa.png)

The **vite.config.js**
![image](https://user-images.githubusercontent.com/17656277/194708889-b3bea3da-95d5-4b39-aba0-ed97a88c9eab.png)

The **postcss.config.js**
![image](https://user-images.githubusercontent.com/17656277/194708852-6d9c4d02-4a42-4d4a-bde0-b70a45dea6b6.png)

Now, if i run `npm run dev` with current postcss config
![image](https://user-images.githubusercontent.com/17656277/194709136-00dd8a28-cdd2-42f7-98e7-0ea19a55590b.png)
Seems like mixin plugin is not working, ok.

I tried removing plugins until I found the config which works:
![image](https://user-images.githubusercontent.com/17656277/194709595-c7b77050-48c7-4026-99b3-ff08db365664.png)
Removing those 2 plugins seemed to be working, output css is what I was expecting
![image](https://user-images.githubusercontent.com/17656277/194709385-17fe2569-403f-466d-80f3-2655049b8ec1.png)

Now
Trying to add only **postcss-simple-vars**:
![image](https://user-images.githubusercontent.com/17656277/194709414-b7ae37ce-c74f-45b6-9c7b-b87cf7188947.png)
Same error as above I think, I don't know if is the @each or the @mixin plugin

Trying to add only **tailwindcss**:
![image](https://user-images.githubusercontent.com/17656277/194709625-d0c618aa-1eff-4ab6-8aaf-2f0978a7c56e.png)
Different error but still @each or @mixin plugin not working well I think

I need all of those plugins and I'm sure if each on of those would be executed one at a time it'll be ok, but in the same process they do this mess.
I'm not sure how plugins are executed but I read that the order matters, so there are any chances that some plugin takes priority or it's been executed asynchronously?

Sorry for the long post, again, here is the repo to reproduce
[https://github.com/giusepetroso/postcss-plugin-issue](https://github.com/giusepetroso/postcss-plugin-issue)

I'm not sure if I should have posted this issue here, or on a specific plugin but actually I can't figure out what is going on so feel free to send me to an other appropriate repo.

Thanks a lot, 
Giuse.

**Comment by ai** [comment]
1. You should not define mixins by `@import`. It is better to use `mixinsDir` and have separated file for each mixin. It will be more reliable and simplify the stack.
2. Order doesn’t matter in PostCSS 8 since plugins run simultaneously.
3. I recommend simplifying your PostCSS plugins. Why do you need `postcss-simple-vars` if you have CSS Custom Properties and `postcss-mixins`? Do you really need `postcss-each`?
4. The best way to use PostCSS is not to replicate Sass approach. It is better to be closer to CSS specs, prefer declarative way over imperative (avoiding writing a code inside CSS).

**Comment by giusepetroso** [comment]
Thanks @ai for the reply.

I have the need to keep some of those plugins due to theming purposes, the example I provided is just a bit of an existing css theme we created.
I didn't know about `mixinsDir` so I'll do a research and I can just simply refactor it.
`postcss-simple-vars` maybe could be ignored, actually we can use just css vars.
For the `postcss-each` unfortunately I don't see another way to do what we can want to do without repeating lots of lines of css, for example, how can I replicate this logic?
![image](https://user-images.githubusercontent.com/17656277/195021542-53c63992-9569-4ccf-be65-4eb282e6cb44.png)
I have the need to easily replicate some styles in various components, based on an array of defined colors.

There is something I can do to keep `postcss-each` and `postcss-mixins` together?

**Comment by ai** [comment]
Can you show me the `--color-palette`? Do you use it somewhere else?

**Comment by giusepetroso** [comment]
Here is the `--color-palette` variable:
![image](https://user-images.githubusercontent.com/17656277/195301312-d825d004-7c2e-49de-a674-53795018537a.png)

I use it just for those kind of loops.

**Comment by ai** [comment]
Why not:

```css
@mixin bb-badge-palette black;
@mixin bb-badge-palette gray;
@mixin bb-badge-palette purple;
@mixin bb-badge-palette aqua;
@mixin bb-badge-palette green;
@mixin bb-badge-palette yellow;
@mixin bb-badge-palette red;
@mixin bb-badge-palette primary;
@mixin bb-badge-palette secondary;
```

**Comment by giusepetroso** [comment]
It's because I need to cycle those colors in lot of components, this is sort of a boilerplate for our projects and at every project we could have a different palette, so it's convenient to define the list of colors once and cycle in every component intead of editing it in 20+ files.

**Comment by ai** [comment]
In this case you will need to keep `postcss-each`.

**Comment by giusepetroso** [comment]
Yeah I think so, but how can I?
I can get rid of `postcss-simple-vars` but still can get it work.
I tried the mixinDir option, and now it compiles whithout errors but it's skipping the `postcss-conditionals`, `tailwindcss/nesting`, `tailwindcss` and `autoprefixer`.
_(I formatted the output to better readability)_
![image](https://user-images.githubusercontent.com/17656277/196479654-089176a1-b122-477a-9dba-8e1e8341fbbc.png)
I think the only way to make it work would be to force some plugins to execute before others, but I don't know if I'm getting the point of why it works async now.

**Comment by giusepetroso** [comment]
Actually I can get rid of it, but issue remains with `tailwindcss/nesting`, `tailwindcss` and `autoprefixer`.

**Comment by ai** [comment]
The big complexity of your case could hide the real issue. We need to simplify stack first.

Maybe you have a simple use case of exact issue with tailwindcss/nesting, tailwindcss and autoprefixer.?

**Comment by giusepetroso** [comment]
I created a repo for this issue https://github.com/giusepetroso/postcss-plugin-issue

**Comment by ai** [comment]
Sorry, it is very big project for debug.

1. Can you keep a simple case of one CSS file with expected output and real output?
2. Can you remove Vue and other non-PostCSS dependencies?

**Comment by giusepetroso** [comment]
Hi, finally I got some time to make the minimal debug project.

Now on the master branch of the same test repo ([https://github.com/giusepetroso/postcss-plugin-issue](https://github.com/giusepetroso/postcss-plugin-issue)) I did some polishing.

I completely removed Vuejs from the compilation.

Doing a recap, this is the postcss.config.js
![image](https://user-images.githubusercontent.com/17656277/204079841-09434645-45c7-4023-9d80-14281b2faba1.png)

this is the vite.config.js
![image](https://user-images.githubusercontent.com/17656277/204079868-ab95accb-4aab-4726-b3a1-5566cd7fad8d.png)

In the /src folder there I put two css files
![image](https://user-images.githubusercontent.com/17656277/204079803-0530eb99-400e-460d-86f8-91600bc1982c.png)
- "actual.css" is the actual compilation running `npm run build`
- "expected.css" is what I expect to have

To make "expected.css" I first ran "npm run build" with this config
![image](https://user-images.githubusercontent.com/17656277/204079945-df681c31-1cde-4ed2-ad59-7131e0887468.png)

then I re-ran "npm run build" (compiling the previous css result) with this config
![image](https://user-images.githubusercontent.com/17656277/204079992-39e28b99-f974-48d2-99f3-e55eda5d3f62.png)

basically splitting the compilation in two phases.

If you need some other explainations ask me.
Thank you for the support.

**Comment by ai** [comment]
[The fix](https://github.com/postcss/postcss-scss/pull/143) was released in `postcss-scss` 4.0.6.

**Comment by giusepetroso** [comment]
I am not getting the point since I've no `postcss-scss` package in my dependencies. 
Can you explain me how can I fix my issue with that?
Thanks.

**Comment by ai** [comment]
Oops. Wrong issue.

(I will look into your bug when I will finish my vacation)

**Comment by giusepetroso** [comment]
No worries man, have a nice vacation :)

**Comment by ai** [comment]
Sorry, I still can’t help here sine the test case is too complicated and big.

Seems like it is better for you to separated Tailwind and other plugins.

**Comment by giusepetroso** [comment]
But the fact is I need to keep Tailwind into the postCSS plugins, it is actually the point, if I remove it there is no issue.

I'm not getting your request, now the project is just the essential css, but Tailwind is part of the problem I cannot exclude it.

Let me understand what do you mean please.

**Comment by ai** [comment]
There is no such thing as plugins order. They all applying to CSS in the same moment with an event system.

To fix your bug, we need to do a lot of work in plugins communication. I will not have time for this.

**Comment by dufusproxy** [comment]
Im facing same issue with a plugin I'm writing to extend tailwind, my plugin works fine (results in css that tailwind would parse correctly if it was the source) but as soo as you add tailwindcss after the plugin it doesn't.
Tailwind doesn't respect the ordering, well that is to say that apparently it gets round ordering by using lightning css which runs first then it passes to postcss. I haven't looked in full detail yet.
It seems they do this to bypass the ordering, the tailwind postcss plugin may still run in order, but the lightning pass has run and has probably already removed any things like @theme, @apply etc before the first postcss plugin s run.

Its frustrating, in my case if i save the postcss result when i dont include tailwindcss as a file. Then i use that file as a src with tailwindcss enabled it all works fine.

But yeah this is a tailwind issue, they have a plugin system via @plugin which accepts js for legacy plugins.
The description made it sound to me like at some point they may deprecate that.
So I decided to write mine so you use css and postcss to convert it before tailwind embracing the css first approach of tw v4, but as others are finding it cant work, so im just hoping I can use a lot of what I have to create a old v tw js plugin.
